Selecting hidden columns fails to open split in Hudi

We should throw column not found exception if we don't support them. 

```
SELECT "$path" FROM hudi.tests.hudi_cow_pt_tbl;

Query 20220928_074646_00017_n54ug failed: Error opening Hudi split file:/var/folders/8s/dkvf18z55lj_9yxhy1n54sph0000gn/T/TrinoTest15358788863231304430/data/hudi_cow_pt_tbl/dt=2021-12-09/hh=10/719c3273-2805-4124-b1ac-e980dada85bf-0_0-27-1215_20220906063435640.parquet (offset=0, length=435204): column type must be REGULAR: $path:string:SYNTHESIZED
io.trino.spi.TrinoException: Error opening Hudi split file:/var/folders/8s/dkvf18z55lj_9yxhy1n54sph0000gn/T/TrinoTest15358788863231304430/data/hudi_cow_pt_tbl/dt=2021-12-09/hh=10/719c3273-2805-4124-b1ac-e980dada85bf-0_0-27-1215_20220906063435640.parquet (offset=0, length=435204): column type must be REGULAR: $path:string:SYNTHESIZED
	at io.trino.plugin.hudi.HudiPageSourceProvider.createPageSource(HudiPageSourceProvider.java:268)
	at io.trino.plugin.hudi.HudiPageSourceProvider.createPageSource(HudiPageSourceProvider.java:165)
	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorPageSourceProvider.createPageSource(ClassLoaderSafeConnectorPageSourceProvider.java:49)
	at io.trino.split.PageSourceManager.createPageSource(PageSourceManager.java:62)
	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:308)
	at io.trino.operator.Driver.processInternal(Driver.java:411)
	at io.trino.operator.Driver.lambda$process$10(Driver.java:314)
	at io.trino.operator.Driver.tryWithLock(Driver.java:706)
	at io.trino.operator.Driver.process(Driver.java:306)
	at io.trino.operator.Driver.processForDuration(Driver.java:277)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:736)
	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:164)
	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:515)
	at io.trino.$gen.Trino_testversion____20220928_074154_36.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IllegalArgumentException: column type must be REGULAR: $path:string:SYNTHESIZED
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:220)
	at io.trino.plugin.hive.parquet.ParquetPageSourceFactory.createParquetReaderColumns(ParquetPageSourceFactory.java:440)
	at io.trino.plugin.hudi.HudiPageSourceProvider.createPageSource(HudiPageSourceProvider.java:227)
	... 16 more
```

Below columns cause this issue
* `$path`
* `$file_size`
* `$file_modified_time`
* `$partition`

--- a/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiPageSource.java
+++ b/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiPageSource.java
@@ -13,36 +13,87 @@
  */
 package io.trino.plugin.hudi;
 
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableMap;
 import io.trino.plugin.hive.HiveColumnHandle;
 import io.trino.spi.Page;
 import io.trino.spi.block.Block;
 import io.trino.spi.block.RunLengthEncodedBlock;
 import io.trino.spi.connector.ConnectorPageSource;
+import org.apache.hadoop.fs.Path;
 
 import java.io.IOException;
 import java.util.List;
 import java.util.Map;
 
+import static io.airlift.slice.Slices.utf8Slice;
 import static io.trino.plugin.base.util.Closables.closeAllSuppress;
+import static io.trino.plugin.hive.HiveColumnHandle.FILE_MODIFIED_TIME_COLUMN_NAME;
+import static io.trino.plugin.hive.HiveColumnHandle.FILE_MODIFIED_TIME_TYPE_SIGNATURE;
+import static io.trino.plugin.hive.HiveColumnHandle.FILE_SIZE_COLUMN_NAME;
+import static io.trino.plugin.hive.HiveColumnHandle.FILE_SIZE_TYPE_SIGNATURE;
+import static io.trino.plugin.hive.HiveColumnHandle.PARTITION_COLUMN_NAME;
+import static io.trino.plugin.hive.HiveColumnHandle.PARTITION_TYPE_SIGNATURE;
+import static io.trino.plugin.hive.HiveColumnHandle.PATH_COLUMN_NAME;
+import static io.trino.plugin.hive.HiveColumnHandle.PATH_TYPE;
+import static io.trino.spi.predicate.Utils.nativeValueToBlock;
+import static io.trino.spi.type.DateTimeEncoding.packDateTimeWithZone;
+import static io.trino.spi.type.TimeZoneKey.UTC_KEY;
 import static java.util.Objects.requireNonNull;
 
 public class HudiPageSource
         implements ConnectorPageSource
 {
-    private final List<HiveColumnHandle> columnHandles;
+    private final Block[] prefilledBlocks;
+    private final int[] delegateIndexes;
     private final ConnectorPageSource dataPageSource;
-    private final Map<String, Block> partitionBlocks;
 
     public HudiPageSource(
+            String partitionName,
             List<HiveColumnHandle> columnHandles,
             Map<String, Block> partitionBlocks,
-            ConnectorPageSource dataPageSource)
+            ConnectorPageSource dataPageSource,
+            Path path,
+            long fileSize,
+            long fileModifiedTime)
     {
-        this.columnHandles = ImmutableList.copyOf(requireNonNull(columnHandles, "columnHandles is null"));
-        this.partitionBlocks = ImmutableMap.copyOf(requireNonNull(partitionBlocks, "partitionBlocks is null"));
+        requireNonNull(columnHandles, "columnHandles is null");
         this.dataPageSource = requireNonNull(dataPageSource, "dataPageSource is null");
+
+        int size = columnHandles.size();
+        this.prefilledBlocks = new Block[size];
+        this.delegateIndexes = new int[size];
+
+        int outputIndex = 0;
+        int delegateIndex = 0;
+
+        for (HiveColumnHandle column : columnHandles) {
+            if (partitionBlocks.containsKey(column.getName())) {
+                Block partitionValue = partitionBlocks.get(column.getName());
+                prefilledBlocks[outputIndex] = partitionValue;
+                delegateIndexes[outputIndex] = -1;
+            }
+            else if (column.getName().equals(PARTITION_COLUMN_NAME)) {
+                prefilledBlocks[outputIndex] = nativeValueToBlock(PARTITION_TYPE_SIGNATURE, utf8Slice(partitionName));
+                delegateIndexes[outputIndex] = -1;
+            }
+            else if (column.getName().equals(PATH_COLUMN_NAME)) {
+                prefilledBlocks[outputIndex] = nativeValueToBlock(PATH_TYPE, utf8Slice(path.toString()));
+                delegateIndexes[outputIndex] = -1;
+            }
+            else if (column.getName().equals(FILE_SIZE_COLUMN_NAME)) {
+                prefilledBlocks[outputIndex] = nativeValueToBlock(FILE_SIZE_TYPE_SIGNATURE, fileSize);
+                delegateIndexes[outputIndex] = -1;
+            }
+            else if (column.getName().equals(FILE_MODIFIED_TIME_COLUMN_NAME)) {
+                long packedTimestamp = packDateTimeWithZone(fileModifiedTime, UTC_KEY);
+                prefilledBlocks[outputIndex] = nativeValueToBlock(FILE_MODIFIED_TIME_TYPE_SIGNATURE, packedTimestamp);
+                delegateIndexes[outputIndex] = -1;
+            }
+            else {
+                delegateIndexes[outputIndex] = delegateIndex;
+                delegateIndex++;
+            }
+            outputIndex++;
+        }
     }
 
     @Override
@@ -72,23 +123,16 @@ public class HudiPageSource
                 return null;
             }
             int positionCount = page.getPositionCount();
-
-            int dataColumnIndex = 0;
-            int columnIndex = 0;
-            Block[] blocksWithPartitionColumns = new Block[columnHandles.size()];
-            for (HiveColumnHandle columnHandle : columnHandles) {
-                if (columnHandle.isPartitionKey()) {
-                    Block partitionValue = partitionBlocks.get(columnHandle.getName());
-                    blocksWithPartitionColumns[columnIndex] = RunLengthEncodedBlock.create(partitionValue, positionCount);
+            Block[] blocks = new Block[prefilledBlocks.length];
+            for (int i = 0; i < prefilledBlocks.length; i++) {
+                if (prefilledBlocks[i] != null) {
+                    blocks[i] = RunLengthEncodedBlock.create(prefilledBlocks[i], positionCount);
                 }
                 else {
-                    blocksWithPartitionColumns[columnIndex] = page.getBlock(dataColumnIndex);
-                    dataColumnIndex++;
+                    blocks[i] = page.getBlock(delegateIndexes[i]);
                 }
-
-                columnIndex++;
             }
-            return new Page(positionCount, blocksWithPartitionColumns);
+            return new Page(positionCount, blocks);
         }
         catch (RuntimeException e) {
             closeAllSuppress(e, this);

--- a/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiPageSourceProvider.java
+++ b/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiPageSourceProvider.java
@@ -115,6 +115,7 @@ import static java.util.Objects.requireNonNull;
 import static java.util.stream.Collectors.toList;
 import static java.util.stream.Collectors.toMap;
 import static java.util.stream.Collectors.toUnmodifiableList;
+import static org.apache.hadoop.hive.common.FileUtils.makePartName;
 
 public class HudiPageSourceProvider
         implements ConnectorPageSourceProvider
@@ -158,16 +159,20 @@ public class HudiPageSourceProvider
         // just send regular columns to create parquet page source
         // for partition columns, separate blocks will be created
         List<HiveColumnHandle> regularColumns = hiveColumns.stream()
-                .filter(columnHandle -> !columnHandle.isPartitionKey())
+                .filter(columnHandle -> !columnHandle.isPartitionKey() && !columnHandle.isHidden())
                 .collect(Collectors.toList());
         TrinoFileSystem fileSystem = fileSystemFactory.create(session);
         TrinoInputFile inputFile = fileSystem.newInputFile(path.toString(), split.getFileSize());
         ConnectorPageSource dataPageSource = createPageSource(session, regularColumns, split, inputFile, dataSourceStats, options, timeZone);
 
         return new HudiPageSource(
+                toPartitionName(split.getPartitionKeys()),
                 hiveColumns,
                 convertPartitionValues(hiveColumns, split.getPartitionKeys()), // create blocks for partition values
-                dataPageSource);
+                dataPageSource,
+                path,
+                split.getFileSize(),
+                split.getFileModifiedTime());
     }
 
     private static ConnectorPageSource createPageSource(
@@ -343,5 +348,16 @@ public class HudiPageSourceProvider
                     format("Can not parse partition value '%s' of type '%s' for partition column '%s'", partitionValue, partitionDataType, partitionColumnName),
                     e);
         }
+    }
+
+    private static String toPartitionName(List<HivePartitionKey> partitions)
+    {
+        ImmutableList.Builder<String> partitionNames = ImmutableList.builderWithExpectedSize(partitions.size());
+        ImmutableList.Builder<String> partitionValues = ImmutableList.builderWithExpectedSize(partitions.size());
+        for (HivePartitionKey partition : partitions) {
+            partitionNames.add(partition.getName());
+            partitionValues.add(partition.getValue());
+        }
+        return makePartName(partitionNames.build(), partitionValues.build());
     }
 }

--- a/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSplit.java
+++ b/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSplit.java
@@ -37,6 +37,7 @@ public class HudiSplit
     private final long start;
     private final long length;
     private final long fileSize;
+    private final long fileModifiedTime;
     private final List<HostAddress> addresses;
     private final TupleDomain<HiveColumnHandle> predicate;
     private final List<HivePartitionKey> partitionKeys;
@@ -48,6 +49,7 @@ public class HudiSplit
             @JsonProperty("start") long start,
             @JsonProperty("length") long length,
             @JsonProperty("fileSize") long fileSize,
+            @JsonProperty("fileModifiedTime") long fileModifiedTime,
             @JsonProperty("addresses") List<HostAddress> addresses,
             @JsonProperty("predicate") TupleDomain<HiveColumnHandle> predicate,
             @JsonProperty("partitionKeys") List<HivePartitionKey> partitionKeys,
@@ -61,6 +63,7 @@ public class HudiSplit
         this.start = start;
         this.length = length;
         this.fileSize = fileSize;
+        this.fileModifiedTime = fileModifiedTime;
         this.addresses = ImmutableList.copyOf(requireNonNull(addresses, "addresses is null"));
         this.predicate = requireNonNull(predicate, "predicate is null");
         this.partitionKeys = ImmutableList.copyOf(requireNonNull(partitionKeys, "partitionKeys is null"));
@@ -88,6 +91,7 @@ public class HudiSplit
                 .put("start", start)
                 .put("length", length)
                 .put("fileSize", fileSize)
+                .put("fileModifiedTime", fileModifiedTime)
                 .buildOrThrow();
     }
 
@@ -123,6 +127,12 @@ public class HudiSplit
     }
 
     @JsonProperty
+    public long getFileModifiedTime()
+    {
+        return fileModifiedTime;
+    }
+
+    @JsonProperty
     public TupleDomain<HiveColumnHandle> getPredicate()
     {
         return predicate;
@@ -142,6 +152,7 @@ public class HudiSplit
                 .addValue(start)
                 .addValue(length)
                 .addValue(fileSize)
+                .addValue(fileModifiedTime)
                 .toString();
     }
 }

--- a/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/split/HudiSplitFactory.java
+++ b/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/split/HudiSplitFactory.java
@@ -61,6 +61,7 @@ public class HudiSplitFactory
                         fileSplit.getStart(),
                         fileSplit.getLength(),
                         fileStatus.getLen(),
+                        fileStatus.getModificationTime(),
                         ImmutableList.of(),
                         hudiTableHandle.getRegularPredicates(),
                         partitionKeys,

--- a/plugin/trino-hudi/src/test/java/io/trino/plugin/hudi/TestHudiSmokeTest.java
+++ b/plugin/trino-hudi/src/test/java/io/trino/plugin/hudi/TestHudiSmokeTest.java
@@ -19,12 +19,17 @@ import io.trino.testing.AbstractTestQueryFramework;
 import io.trino.testing.QueryRunner;
 import org.testng.annotations.Test;
 
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.ZonedDateTime;
+
 import static io.trino.plugin.hudi.HudiQueryRunner.createHudiQueryRunner;
 import static io.trino.plugin.hudi.testing.ResourceHudiTablesInitializer.TestingTable.HUDI_COW_PT_TBL;
 import static io.trino.plugin.hudi.testing.ResourceHudiTablesInitializer.TestingTable.HUDI_NON_PART_COW;
 import static io.trino.plugin.hudi.testing.ResourceHudiTablesInitializer.TestingTable.STOCK_TICKS_COW;
 import static io.trino.plugin.hudi.testing.ResourceHudiTablesInitializer.TestingTable.STOCK_TICKS_MOR;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.testng.Assert.assertEquals;
 
 public class TestHudiSmokeTest
         extends AbstractTestQueryFramework
@@ -112,5 +117,45 @@ public class TestHudiSmokeTest
                         "   location = \\E'.*/hudi_cow_pt_tbl',\n\\Q" +
                         "   partitioned_by = ARRAY['dt','hh']\n" +
                         ")");
+    }
+
+    @Test
+    public void testPathColumn()
+    {
+        String path = (String) computeScalar("SELECT \"$path\" FROM " + HUDI_COW_PT_TBL + " WHERE id = 1");
+        assertThat(toPath(path)).exists();
+    }
+
+    @Test
+    public void testFileSizeColumn()
+            throws Exception
+    {
+        String path = (String) computeScalar("SELECT \"$path\" FROM " + HUDI_COW_PT_TBL + " WHERE id = 1");
+        long fileSize = (long) computeScalar("SELECT \"$file_size\" FROM " + HUDI_COW_PT_TBL + " WHERE id = 1");
+        assertEquals(fileSize, Files.size(toPath(path)));
+    }
+
+    @Test
+    public void testFileModifiedColumn()
+            throws Exception
+    {
+        String path = (String) computeScalar("SELECT \"$path\" FROM " + HUDI_COW_PT_TBL + " WHERE id = 1");
+        ZonedDateTime fileModifiedTime = (ZonedDateTime) computeScalar("SELECT \"$file_modified_time\" FROM " + HUDI_COW_PT_TBL + " WHERE id = 1");
+        assertEquals(fileModifiedTime.toInstant().toEpochMilli(), Files.getLastModifiedTime(toPath(path)).toInstant().toEpochMilli());
+    }
+
+    @Test
+    public void testPartitionColumn()
+    {
+        assertQuery("SELECT \"$partition\" FROM " + HUDI_COW_PT_TBL + " WHERE id = 1", "VALUES 'dt=2021-12-09/hh=10'");
+        assertQuery("SELECT \"$partition\" FROM " + HUDI_COW_PT_TBL + " WHERE id = 2", "VALUES 'dt=2021-12-09/hh=11'");
+
+        assertQueryFails("SELECT \"$partition\" FROM " + HUDI_NON_PART_COW, ".* Column '\\$partition' cannot be resolved");
+    }
+
+    private static Path toPath(String path)
+    {
+        // Remove leading 'file:' because $path column returns 'file:/path-to-file' in case of local file system
+        return Path.of(path.replaceFirst("^file:", ""));
     }
 }

--- a/plugin/trino-hudi/src/test/java/io/trino/plugin/hudi/TestHudiSmokeTest.java
+++ b/plugin/trino-hudi/src/test/java/io/trino/plugin/hudi/TestHudiSmokeTest.java
@@ -37,7 +37,7 @@ public class TestHudiSmokeTest
     }
 
     @Test
-    public void readNonPartitionedTable()
+    public void testReadNonPartitionedTable()
     {
         assertQuery(
                 "SELECT rowid, name FROM " + HUDI_NON_PART_COW,
@@ -45,7 +45,7 @@ public class TestHudiSmokeTest
     }
 
     @Test
-    public void readPartitionedTables()
+    public void testReadPartitionedTables()
     {
         assertQuery("SELECT symbol, max(ts) FROM " + STOCK_TICKS_COW + " GROUP BY symbol HAVING symbol = 'GOOG'",
                 "SELECT * FROM VALUES ('GOOG', '2018-08-31 10:59:00')");

